### PR TITLE
Delete local execution scopes

### PR DESCRIPTION
**Pull request by chengduoZH**
Add scope_buffered_monitor to monitored the local execution scopes.

### CODE REVIEW

**Review by sneaxiy — COMMENTED** [review]
As a user, using a flag to control the memory leak sizes has no differences with setting `exec_strategy.num_iteration_per_drop_scope = xxx`. It is also a burden for users.

**Review by sneaxiy — COMMENTED** [review]
BTW, for large models which occupy lots of GPU memory, limitation of 1GB is even too large.

**Comment by chengduoZH** [comment]
> As a user, using a flag to control the memory leak sizes has no differences with setting exec_strategy.num_iteration_per_drop_scope = xxx. It is also a burden for users.

It is different in fact,  this flag focuses on the sub-scope of local execution scope, but not variables of local execution scope, those variables can be collected by GC now. For some models, if the operators in this graph(converted from a model) are run in different places, there have sub-scope in local execution scope, if not, there may not have sub-scope in local execution scope.

File: paddle/fluid/framework/details/scope_buffered_monitor.cc [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+  for (size_t scope_id = 0; scope_id < local_exec_scopes_.size(); ++scope_id) {
+    for (auto &scope : post_local_exec_scopes_.at(scope_id)) {
+      if (!pre_local_exec_scopes_.at(scope_id).count(scope)) {
+        incr_local_exec_scopes.at(scope_id).insert(scope);
+      }
+    }
+
+    if (incr_local_exec_scopes.at(scope_id).size() &&
+        FLAGS_local_exe_sub_scope_limit > 0) {
+      LOG_FIRST_N(WARNING, 1)
```
Comment (sneaxiy) [inline paddle/fluid/framework/details/scope_buffered_monitor.cc]: Do we need to add this warning to users? Users may not know what it means.

File: paddle/fluid/platform/flags.cc [hunk 2]
```diff
@@ -427,3 +427,17 @@ DEFINE_uint64(reallocate_gpu_memory_in_mb, 0ul,
...
+ * Scope related FLAG
+ * Name: local_exe_sub_scope_limit
+ * Since Version: 1.6.0
+ * Value Range: double, default=256 (MB)
+ * Example:
+ * Note:
+ */
+DEFINE_double(local_exe_sub_scope_limit, 256.0,  // MBytes
+              "The memory up limit of sub-scopes of local execution scope for "
+              "each GPUPlace. If you don't need to limit the memory, "
```
Comment (sneaxiy) [inline paddle/fluid/platform/flags.cc]: How about "each CUDAPlace"?

**Review by sneaxiy — DISMISSED** [review]
LGTM for `ScopeBufferedMonitor`.

**Comment by lanxianghit** [comment]
FLAG的用户使用说明文档在哪里？

**Comment by chengduoZH** [comment]
@lanxianghit 
> FLAG的用户使用说明文档在哪里？

这里 https://github.com/PaddlePaddle/Paddle/pull/19749/files#diff-441c4a3926cd9492a8c825c8e718e9b3R431

File: paddle/fluid/platform/flags.cc (hunk 2 above)
Comment (chengduoZH) [reply paddle/fluid/platform/flags.cc]: Ok, I will fix this in next PR.

File: paddle/fluid/framework/details/scope_buffered_monitor.cc (hunk 1 above)
Comment (chengduoZH) [reply paddle/fluid/framework/details/scope_buffered_monitor.cc]: Ok, I will fix this in next PR.

**Review by lanxianghit — APPROVED** [review]
LGTM

File: paddle/fluid/platform/flags.cc (hunk 2 above)
Comment (chengduoZH) [reply paddle/fluid/platform/flags.cc]: Done

File: paddle/fluid/framework/details/scope_buffered_monitor.cc (hunk 1 above)
Comment (chengduoZH) [reply paddle/fluid/framework/details/scope_buffered_monitor.cc]: Done.

**Pull request merged**